### DISCUSSION TITLE
specifying java version for smithy generator

### DISCUSSION THREAD

**Discussion by micro-jumbo** [aws/aws-pdk#706] [Q&A]
Hi,

I have a model defined in smithy and I'm trying to build it using github actions. On my local machine everything works fine, but when run as an actions I get the following error:
```
my-api-model: FAILURE: Build failed with an exception.
my-api-model: * What went wrong:
my-api-model: Execution failed for task ':smithyBuildJar'.
my-api-model: > Could not resolve all files for configuration ':runtimeClasspath'.
my-api-model:    > Could not resolve software.amazon.smithy.typescript:smithy-typescript-codegen:0.19.0.
my-api-model:      Required by:
my-api-model:          project :
my-api-model:       > No matching variant of software.amazon.smithy.typescript:smithy-typescript-codegen:0.19.0 was found. The consumer was configured to find a library for use during runtime, compatible with Java 11, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally but:
my-api-model:           - Variant 'apiElements' capability software.amazon.smithy.typescript:smithy-typescript-codegen:0.19.0 declares a library, packaged as a jar, and its dependencies declared externally:
my-api-model:               - Incompatible because this component declares a component for use during compile-time, compatible with Java 17 and the consumer needed a component for use during runtime, compatible with Java 11
my-api-model:               - Other compatible attribute:
my-api-model:                   - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
```

I'm guessing that's because github action runs using Java 11, but smithy libraries are built for Java 17. Where can I set the java version to be using by the generator?

**Comment by cogwirrel** [comment]
Hey Cyprian!

The Type Safe API code generation should be compatible with Java 11+ 😄 

It looks here like you're using Smithy codegen... I'm not super familiar with that just yet as we're still figuring out how we start to add that support natively in Type Safe API, but I wonder if it's the Gradle JDK version that needs to be specified here?

Perhaps you could try adding the following to your `model/build.gradle` file (just override it to test since it's read-only and managed by projen):

```gradle
java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(11))
    }
}
```

Happy to have a chat if it helps - feel free to ping me on the https://cdk.dev/ slack!

Cheers,
Jack

**Reply by micro-jumbo** [reply]
Hi Jack,

I think the problem is that Smithy codegen libraries are built using Java 17, but GitHub workflows are using Java 11 by default. When I added this step to `build` workflow:
```
      - name: Setup java
        uses: actions/setup-java@v4
        with:
          distribution: 'corretto'
          java-version: '17'
```
(has to be added _after_ "actions/setup-node@v4") it started working. I added it manually to `$projectRoot/.github/workflows/build.yml`, but I'd like to know how can I add it using pdk/projen?

Thanks,
Cyprian

**Reply by cogwirrel** [reply]
Ahh I see, makes sense! I've not done much github workflow management with projen so there may be a better way than this, but I found that this seemed to work ok:

```ts
project.github?.tryFindWorkflow('build')?.file?.patch(JsonPatch.add('/jobs/build/steps/3', {
  name: 'setup java',
  uses: 'actions/setup-java@v4',
  with: {
    distribution: 'corretto',
    'java-version': '17',
  },
}));
```

Where `3` is just an index I picked arbitrarily to insert that step.

Hope this helps! 😄
